### PR TITLE
Track queue size of bounded queues

### PR DESCRIPTION
**Pull request by by-dam**
tentative to address https://github.com/functional-streams-for-scala/fs2/issues/1448

### CODE REVIEW

File: core/shared/src/main/scala/fs2/concurrent/PubSub.scala [hunk 1]
```diff
@@ -436,6 +436,36 @@ private[fs2] object PubSub {
           strategy.unsubscribe(selector, state)
       }
 
+    def withSize[A, S](strategy: PubSub.Strategy[A, Chunk[A], S, Int])
+      : PubSub.Strategy[A, Chunk[A], (S, Int), Int] =
+      new PubSub.Strategy[A, Chunk[A], (S, Int), Int] {
```
Comment (by-dam) [inline core/shared/src/main/scala/fs2/concurrent/PubSub.scala]: strategy that tracks the queue size along with the state `S`

**Comment by mpilquist** [comment]
@by-dam Looks like we need similar changes in the experimental module too.

**Comment by mpilquist** [comment]
@by-dam I changed the approach a bit, introducing a new `SizedQueue` class. Hope that's okay.

**Comment by SystemFw** [comment]
> I changed the approach a bit, introducing a new SizedQueue class. Hope that's okay.

Fwiw, I was also a bit unsure whether the size tracking belonged to Queue or PubSub, but couldn't really decide either way

**Comment by mpilquist** [comment]
@SystemFw Yeah, when I getting this PR ready for merge, I noticed `circularBuffer` needed a similar treatment b/c each enqueue checked `q.size`, and the original strategy from this PR didn't help in that case. Are you OK with merging this new version?

**Comment by btlines** [comment]
> I changed the approach a bit, introducing a new SizedQueue class. Hope that's okay.

@mpilquist Sure, thanks for looking at it

**Comment by SystemFw** [comment]
Yes, let's merge this :)

**Comment by by-dam** [comment]
>  I changed the approach a bit, introducing a new SizedQueue class. Hope that's okay.

@mpilquist Sure, thanks for looking at it

**Pull request merged**